returns/interfaces/failable.py

```diff
@@ -0,0 +1,166 @@
+from returns.interfaces.rescuable import RescuableN
+from returns.primitives.asserts import assert_equal
+from returns.primitives.hkt import KindN
+from returns.primitives.laws import (
+    Law,
+    Law3,
+    Lawful,
+    LawSpecDef,
+    law_definition,
+)
+
+_FirstType = TypeVar('_FirstType')
+_SecondType = TypeVar('_SecondType')
+_ThirdType = TypeVar('_ThirdType')
+_UpdatedType = TypeVar('_UpdatedType')
+
+_DiverseFailableType = TypeVar('_DiverseFailableType', bound='DiverseFailableN')
+
+# Used in laws:
+_NewFirstType = TypeVar('_NewFirstType')
+
+
+@final
+class _FailableLawSpec(LawSpecDef):
+    """
+    Failable laws.
+
+    We need to be sure that ``.rescue`` won't rescue success types.
+    """
+
+    @law_definition
+    def rescue_short_circuit_law(
+        raw_value: _FirstType,
+        container: 'FailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_SecondType],
+            KindN['FailableN', _FirstType, _NewFirstType, _ThirdType],
+        ],
+    ) -> None:
+        """Ensures that you cannot rescue a success."""
+        assert_equal(
+            container.from_value(raw_value),
+            container.from_value(raw_value).rescue(function),
+        )
+
+
+@final
+class _DiverseFailableLawSpec(LawSpecDef):
+    """
+    Diverse Failable laws.
+
+    We need to be sure that ``.map``, ``.bind`` and ``.alt``
+    works correctly for both success and failure types.
+    """
+
+    @law_definition
+    def map_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot map a failure."""
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).map(function),
+        )
+
+    @law_definition
+    def bind_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_FirstType],
+            KindN['DiverseFailableN', _NewFirstType, _SecondType, _ThirdType],
+        ],
+    ) -> None:
+        """
+        Ensures that you cannot bind a failure.
+
+        See: https://wiki.haskell.org/Typeclassopedia#MonadFail
+        """
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).bind(function),
+        )
+
+    @law_definition
+    def alt_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_SecondType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot alt a success."""
+        assert_equal(
+            container.from_value(raw_value),
+            container.from_value(raw_value).alt(function),
+        )
+
+
+class FailableN(
```

Comment:
We also tend to sctructure interface files differently: 
- LawsSpec
- Its interface
- Another LawsSpec
- Another Interface